src/OrchardCore/OrchardCore.Modules/PoweredByMiddleware.cs

```diff
@@ -0,0 +1,71 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+
+namespace OrchardCore.Modules
+{
+    /// <summary>
+    /// Adds the X-Powered-By header with values OrchardCore.
+    /// </summary>
+    internal class PoweredByMiddleware
+    {
+        private readonly RequestDelegate _next;
+        private readonly IPoweredByMiddlewareOptions _options;
+
+        public PoweredByMiddleware(RequestDelegate next, IPoweredByMiddlewareOptions options)
+        {
+            _next = next;
+            _options = options;
+        }
+
+        public async Task Invoke(HttpContext httpContext)
+        {
+            if (_options.Enabled)
+                AddHeader(httpContext, _options.HeaderName, _options.HeaderValue);
+            await _next.Invoke(httpContext);
+        }
+
+        /// <summary>
+        /// Adds a delegate to responses OnStarting event in order to add
+        /// a header with <paramref name="headerName"/> and value <paramref name="headerValue"/>
+        /// </summary>
+        /// <param name="httpContext">The Http Context</param>
+        /// <param name="headerName">Header name</param>
+        /// <param name="headerValue">Header value</param>
+        private static void AddHeader(HttpContext httpContext, string headerName, string headerValue)
+        {
+            httpContext.Response.OnStarting(() =>
+            {
+                var headers = httpContext.Response.Headers;
+
+                if (!headers.ContainsKey(headerName))
```

Comment:
Can it be written ` headers[headerName] = headerValue;` without any key lookup?

Reply:
Well, i thought that if there's another middleware inserting the header first, we should append our Powered By. Katana for example it's doing it adding a new header either way and it results on two Powered-By headers.
Anyway its just works, and yours its a KISS solution, so I agree on that.